quinn-proto/src/connection.rs

```diff
@@ -909,11 +909,13 @@ where
                 return;
             }
             DataSent { .. } => {
-                self.events.push_back(Event::StreamFinished {
-                    stream: stream_id,
-                    stop_reason,
-                });
                 // No need to hold on to the stop_reason since it's propagated above
+                if stopped {
```

Comment:
This seems to clearly belong in a separate commit.

Reply:
Is it correct?

Reply:
I think @Ralith knows better than I do.

Reply:
This doesn't impact the protocol state machine, but changes the quinn-proto API behavior such that if you explicitly reset a stream that you're already finishing, no `StreamFinished` event is yielded. This does not affect quinn, as the interface is constructed to statically forbid that.

I can see the case for the change; `StreamFinished` normally indicates that either the stream was successfully delivered in full or the peer stopped it. I'm a bit concerned that in the hypothetical world where you're finishing and resetting a stream concurrently this doesn't give feedback about the status of the finish operation, but as we don't currently permit that in quinn and a hypothetical quinn-proto user won't necessarily need quinn-proto to communicate that for them, I think the change is reasonable as written--though I agree it belongs in an isolated commit.

Reply:
nit: Strictly speaking, the commit message should read "locally reset."

Reply:
I am going to revert this.  Guaranteeing that every successful call to `Connnection::finish` results in a `StreamFinished` event (unless the connection is lost) makes API design easier, and avoids difficult-to-debug hangs.

Reply:
It might be worth adding (in a future PR) a flag in `StreamFinished` which indicates that the finish "completed" due to a local reset, then, since as it stands it's indistinguishable from success.